### ISSUE TITLE
Why not use low precision matmul for reverse embedding in gemma model

### ISSUE THREAD

**Issue by lingzhi98** [keras-team/keras-nlp#1542]
**Describe the bug**
Reverse embedding layer of gemma model use fp32 as reverse_dtype (https://github.com/keras-team/keras-nlp/blob/master/keras_nlp/layers/modeling/reversible_embedding.py#L127), but I find huggingface pytorch implementation (https://github.com/huggingface/transformers/blob/main/src/transformers/models/gemma/modeling_gemma.py#L1113) only cast output to fp32 but not include inputs.

**Expected behavior**
Wish to behave same as huggingface if there exist no impact on accuracy.

**Comment by lingzhi98** [comment]
Thanks. My understanding is that use float32 here to ensure the stability of training, is it necessary for inference also?

**Comment by mattdangerw** [comment]
@lingzhi98 thanks for filing and timely. We actually just hit this as it was causing some numerical differences at half precision for llama 2. I do wonder if we could flip this safely, the best default behavior would be to follow the default dtype strategy whatever that is. cc @tirthasheshpatel who was working on this for llama.

I'll try to check out if this is a safe change to make (particularly with Gemma). As long as we make sure to cast to float32 before any output softmax during sampling, seems like we should be ok.

**Comment by mattdangerw** [comment]
Here's the fix, will push this out next week https://github.com/keras-team/keras-nlp/pull/1548
